sample/redo/lib/main.dart

```diff
@@ -140,125 +142,122 @@ class _MyHomePageState extends State<MyHomePage> {
     return data == null ? null : Todo.fromJson(data);
   }
 
-  Future<void> _write(dynamic tx, int id, Todo todo) {
-    return Future.error('Not implemented');
-    // return _replicache.put(addPrefix(key), todo.toJson());
+  static Future<void> _write(WriteTransaction tx, Todo todo) {
+    final key = addPrefix(todo.id);
+    return tx.put(key, todo);
+  }
+
+  static Future<bool> _del(WriteTransaction tx, int id) =>
+      tx.del(addPrefix(id));
+
+  void _selectListId(int listId) {
+    setState(() {
+      _selectedListId = listId;
+    });
   }
 
-  Future<void> _del(dynamic tx, int id) {
+  Future<void> _dropDatabase() async {
     throw UnimplementedError();
-    // return _replicache.del(addPrefix(key));
+    // Navigator.pop(context);
+    // var items = await _replicache.scan(prefix: prefix);
+    // for (final ScanItem item in items) {
+    //   await _replicache.del(item.key);
+    // }
+    // await _init();
   }
 
-  Future<void> _handleDone(int id, bool complete) async {
-    // TODO(arv): This should be mutate
-    _replicache.query((tx) async {
-      var todo = await _read(tx, id);
+  Mutator<void, Map<String, dynamic>> _todoCreateMutator;
+  Mutator<void, Map<String, dynamic>> _todoDoneMutator;
+  Mutator<void, Map<String, dynamic>> _todoRemoveMutator;
+  Mutator<void, Map<String, dynamic>> _todoReorderMutator;
+
+  _registerMutations() {
+    _todoCreateMutator = _replicache.register('todo-create',
+        (tx, Map<String, dynamic> args) async {
+      int id = args['id'];
+      String text = args['text'];
+      int listId = args['listId'];
+      Iterable<Todo> todos = await todosInListFromTx(tx, listId);
+      int index = todos.isEmpty ? 0 : todos.length;
+
+      final order = getNewOrder(index, todos);
+      await _write(tx, Todo(id, listId, text, false, order));
+    });
+
+    _todoDoneMutator = _replicache.register('todo-done',
+        (tx, Map<String, dynamic> args) async {
+      int id = args['id'];
+      bool complete = args['complete'];
+      final todo = await _read(tx, id);
       if (todo == null) {
         return;
       }
       todo.complete = complete;
-      _write(tx, id, todo);
+      await _write(tx, todo);
     });
-  }
 
-  List<Todo> _activeTodos() {
-    final todos =
-        _allTodos.where((todo) => todo.listId == _selectedListId).toList();
-    todos.sort((t1, t2) => (t1.order - t2.order).sign.toInt());
-    return todos;
-  }
+    _todoRemoveMutator = _replicache.register('todo-remove', (tx, args) async {
+      int id = args['id'];
+      await _del(tx, id);
+    });
 
-  Future<void> _handleReorder(int oldIndex, int newIndex) async {
-    final todos = _activeTodos();
-    int id = todos[oldIndex].id;
-    double order = _getNewOrder(newIndex);
-    // TODO(arv): Should be mutate
-    _replicache.query((tx) async {
-      var todo = await _read(tx, id);
+    _todoReorderMutator =
+        _replicache.register('todo-reorder', (tx, args) async {
+      int listId = args['listId'];
```

Comment:
This part of the SDK bugs me. I wish that the mutator args could be strongly typed and the marshalling and unmarshalling from storage was handled by Replicache internally. We can do that right? You'd just have to have the restriction that all the args used are "json compatible", or else have their own encoders?

Reply:
`json.encode` invokes `toJson()` if present. The other direction has no automatic way.


We can do that but it requires that someone creates a class for the arguments which gets pretty verbose.

Reply:
Argh, the dart type system.

Reply:
I feel like classic example of a feature nobody was asking for.